sdks/java/core/src/main/java/org/apache/beam/sdk/options/PipelineOptionsFactory.java

```diff
@@ -1579,7 +1538,7 @@ public int compare(Method o1, Method o2) {
   private static <T extends PipelineOptions> Map<String, Object> parseObjects(
       Class<T> klass, ListMultimap<String, String> options, boolean strictParsing) {
     Map<String, Method> propertyNamesToGetters = Maps.newHashMap();
-    PipelineOptionsFactory.validateWellFormed(klass, REGISTERED_OPTIONS);
+    CACHE.get().validateWellFormed(klass);
```

Comment:
You should move all the private static methods that refer to the inner contents of the cache into the Cache so that your working over a consistent set of values contained in the cache otherwise your not threadsafe since the contents of the CACHE may change from this line down to line 1581 where we get the CACHE again if resetCache is called in between.

Thus effectively making a non static version of the PipelineOptionsFactory.

Reply:
Tried to refactor a bit what was possible - not public - and without impacting the whole codebase in this PR (can require a big refactoring to have a single singleton in the whole codebase but this would impact a lot of classes and would break all other PR, I think short terms we are ok with current version).

Reply:
Thinking about this more and seeing how the code has progressed, it really seems like we should make PipelineOptionsFactory not a singleton and something that can be constructed and passed around. This would solve the issue where we are leaking stuff because users could construct/pass around a PipelineOptionsFactory.

All the existing static methods would become non-static. We would store a DEFAULT instance within the PipelineOptionsFactory and create new static methods that interact with the DEFAULT instance. This also seems like it would be a lot less work to mark all the existing methods to be non-static as well then to magically call the cache in a bunch of static methods.